lib/python/pyflyby/_modules.py

```diff
@@ -325,37 +328,98 @@ def exports(self):
         """
         Get symbols exported by this module.
 
-        Note that this requires involves actually importing this module, which
-        may have side effects.  (TODO: rewrite to avoid this?)
+        Note that this will not recognize symbols that are dynamically
+        introduced to the module's namespace or __all__ list.
 
         :rtype:
           `ImportSet` or ``None``
         :return:
           Exports, or ``None`` if nothing exported.
         """
         from pyflyby._importclns import ImportStatement, ImportSet
-        module = self.module
+
+        filename = None
         try:
-            members = module.__all__
-        except AttributeError:
-            members = dir(module)
-            # Filter by non-private.
-            members = [n for n in members if not n.startswith("_")]
-            # Filter by definition in the module.
-            def from_this_module(name):
-                # TODO: could do this more robustly by parsing the AST and
-                # looking for STOREs (definitions/assignments/etc).
-                x = getattr(module, name)
-                m = getattr(x, "__module__", None)
-                if not m:
-                    return False
-                return DottedIdentifier(m).startswith(self.name)
-            members = [n for n in members if from_this_module(n)]
-        else:
+            filename = self.filename
+        except:
+            pass
+        if not filename or not filename.exists:
+            # Try to load the module to get the filename
+            filename = Filename(self.module.__file__)
+        text = FileText(filename)
+
+        ast_mod = ast.parse(str(text), str(filename)).body
+
+        # First, add members that are explicitly defined in the module
+        def member_from_node(node):
+            extractors = {
+                # Top-level assignments (as opposed to member assignments
+                # whose targets are of type ast.Attribute).
+                ast.Assign:      lambda x: [t.id for t in x.targets
+                                            if isinstance(t, ast.Name)],
+                ast.ClassDef:    lambda x: [x.name],
+                ast.FunctionDef: lambda x: [x.name],
+            }
+            if isinstance(node, tuple(extractors.keys())):
+                return extractors[type(node)](node)
+            else:
+                return []
```

Comment:
This does not look to me like  a closure, does it capture anything ? 
If not, move it out as a utility, or private class method ?